aip/0162.md

```diff
@@ -0,0 +1,229 @@
+---
+aip:
+  id: 162
+  state: draft
+  created: 2019-09-17
+permalink: /162
+redirect_from:
+  - /0162
+---
+
+# Resource Revisions
+
+Some APIs need to have resources with a revision history, where users can
+reason about the state of the resource over time. There are several reasons for
+this:
+
+- Users may want to be able to roll back to a previous revision, or diff
+  against a previous revision.
+- An API may create data which is derived in some way from a resource at a
+  given point in time. In these cases, it may be desirable to snapshot the
+  resource for reference later.
+
+## Guidance
+
+APIs **may** store a revision history for a resource if it is useful to users.
+
+APIs implementing resources with a revision history **must** provide a
+`revision_id` field on the resource:
+
+```proto
+message Book {
+  // The name of the book.
+  string name = 1;
+
+  // Other fields…
+
+  // The revision ID of the book.
+  // This is automatically changed whenever the contents of the book
+  // change in any way.
+  string revision_id = 99 [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+
+- The `revision_id` field **should** be a string and contain a short,
+  automatically-generated hash. A good rule of thumb is the last eight
+  characters of a UUID4.
+- The field **must** document when new revisions are created.
+
+### Referencing revisions
+
+When it is necessary to refer to a specific revision of a resource, APIs
+**must** use the following syntax: `{resource_name}@{revision_id}`. For
+example:
+
+    publishers/123/books/les-miserables@c7cfa2a8
```

Comment:
This matches syntax used in git, perforce, and elsewhere, which makes it attractive.
I checked, and `@` does _not_ require URL-encoding.

Reply:
We should be careful here. In GCS, for example, you might have a file with an @ sign in the name. It might be difficult to distinguish between an object called "foo@bar" and an object called "foo" at revision "bar".

Reply:
That is a really good point. For normal resource names, Hong wants to limit to DNS characters, but GCS is a really good example of a case where this may not hold up well.

It may still be a good system, but we would need a fallback for that use case if it were to arise.

Reply:
Sold, once I figure out what the alternative is (probably a second field).

Reply:
FWIW, the GCS community tends to use `#` to indicate version.

Reply:
@sethvargo I would love to follow precedent here, but `#` has special meaning in URIs and probably can not work.

Reply:
I added an appendix for this, intentionally avoiding being particularly opinionated.

Reply:
Since there's no motivation or design doc here...

Why invent a new collection syntax for revisions rather than just use a normal nested collection?